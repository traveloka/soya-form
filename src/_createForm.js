import React from 'react';
import PropTypes from 'prop-types';
import hoistStatics from 'hoist-non-react-statics';
import {
  addErrorMessages,
  clear,
  clearErrorMessages,
  mergeFields,
  setDefaultValue,
  setDefaultValues,
  setErrorMessages,
  setFormEnabledState,
  setFormIsSubmittingState,
  setValue,
  setValues,
  initField,
} from './_actions';
import { CONTEXT_NAME } from './_constants';
import { formShape } from './_propTypes';
import {
  getDisplayName,
  throwAndLogError,
} from './_utils';

export const __createForm = (fields, fieldNames) => (formId, dispatch, preventRemoveField) => {
  const __fields = fields || {};
  const __fieldNames = fieldNames || [];
  const actionCreators = {
    lockSubmission: () => setFormIsSubmittingState(formId, true),
    unlockSubmission: () => setFormIsSubmittingState(formId, false),
    disable: () => setFormEnabledState(formId, false),
    enable: () => setFormEnabledState(formId, true),
    disableField(fieldName) {
      return this.disableFields([fieldName]);
    },
    enableField(fieldName) {
      return this.enableFields([fieldName]);
    },
    disableFields: fieldNames => mergeFields(formId, fieldNames.map(fieldName => ({
      fieldName,
      object: { isEnabled: false },
    }))),
    enableFields: fieldNames => mergeFields(formId, fieldNames.map(fieldName => ({
      fieldName,
      object: { isEnabled: true },
    }))),
    setValue: (fieldName, value) => setValue(formId, fieldName, value),
    setValues: values => setValues(formId, values),
    setDefaultValue: (fieldName, value) => setDefaultValue(formId, fieldName, value),
    setDefaultValues: values => setDefaultValues(formId, values),
    setErrors: (fieldName, errorMessages) => setErrorMessages(formId, fieldName, errorMessages),
    addErrors: errorMessages => addErrorMessages(formId, errorMessages),
    clearForm: () => clear(formId),
    clearErrors: fieldNames => clearErrorMessages(formId, fieldNames),
  };

  return {
    ...Object.keys(actionCreators).reduce((methods, name) => ({
      ...methods,
      [name]: (...args) => dispatch(actionCreators[name](...args)),
    }), {}),
    getFormId: () => formId,
    regField: (fieldNames, validateAll) => {
      dispatch(initField(formId, fieldNames));
      __fields[fieldNames] = { validateAll };
      __fieldNames.push(fieldNames);
    },
    unregField: fieldNames => {
      if (!preventRemoveField) {
        delete __fields[fieldNames];
        const index = __fieldNames.findIndex(
          __fieldNames => (
            __fieldNames.length === fieldNames.length &&
            __fieldNames.every((name, i) => name === fieldNames[i])
          ));

        if (index !== -1) {
          __fieldNames.splice(index, 1);
        }
      }
    },
    validateAll: () => {
      const promises = Object.keys(__fields).map(fieldName => __fields[fieldName].validateAll());
      return Promise.all(promises)
        .then(results => {
          let values = {};
          if (
            results.length > 0 &&
            results[0].name.length > 0 &&
            typeof results[0].name[0] === 'number'
          ) {
            values = [];
          }
          return results.reduce((newResult, result) => ({
            isValid: newResult.isValid && result.isValid,
            values: result.name.reduce((value, name, index) => {
              if (index === result.name.length - 1) {
                value[name] = result.value;
                return newResult.values;
              }
              if (!value.hasOwnProperty(name)) {
                value[name] = typeof result.name[index + 1] === 'number' ? [] : {};
              }
              return value[name];
            }, newResult.values),
          }), {
            isValid: true,
            values,
          });
        })
        .catch(throwAndLogError);
    },
    submit(submitFunc, validationFunc) {
      dispatch(clearErrorMessages(formId, __fieldNames));
      return this.validateAll()
        .then(result => {
          if (!result.isValid || !validationFunc) {
            submitFunc(result);
            return;
          }

          return Promise.resolve(validationFunc(result.values))
            .then(validationResult => {
              if (typeof validationResult === 'object' && !validationResult.isValid) {
                dispatch(addErrorMessages(formId, validationResult.errorMessages));
                result.isValid = false;
              }
              submitFunc(result);
            })
            .catch(error => {
              result.isValid = false;
              submitFunc(result);
              throwAndLogError(error);
            });
        })
        .catch(error => {
          throwAndLogError(error);
        });
    },
  };
};

export const _createForm = __createForm();

export default (formId, preventRemoveField) => Component => hoistStatics(class extends React.Component {
  static displayName = getDisplayName('CreateForm', Component);

  static childContextTypes = {
    [CONTEXT_NAME]: formShape.isRequired,
  };

  static propTypes = {
    dispatch: PropTypes.func.isRequired,
    init: PropTypes.func.isRequired,
    formId: PropTypes.string,
  };

  getChildContext = () => ({
    [CONTEXT_NAME]: this.__form,
  });

  constructor(props, context) {
    super(props, context);
    let newFormId = formId || props.formId;
    if (typeof newFormId === 'function') {
      newFormId = newFormId(props);
    }
    this.__form = _createForm(newFormId, props.dispatch, preventRemoveField);
  }

  componentWillMount() {
    this.props.init();
  }

  render() {
    const props = { ...this.props };
    delete props.dispatch;
    delete props.formId;
    props.form = this.__form;
    return <Component {...props} />;
  }
}, Component);
