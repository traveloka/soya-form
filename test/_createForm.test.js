import React from 'react';
import TestUtils from 'react-dom/test-utils';
import _createForm, { __createForm } from '../src/_createForm';

const formId = 'simpleForm';
const dispatch = action => action;

describe('form container', () => {
  class Form extends React.Component {
    render() {
      return <div />;
    }
  }

  it('should receive the form in the props', () => {
    const FormContainer = _createForm()(Form);
    const tree = TestUtils.renderIntoDocument(
      <FormContainer
        dispatch={dispatch}
        formId={formId}
        init={jest.fn()}
      />
    );
    const form = TestUtils.findRenderedComponentWithType(tree, Form);
    expect(form.props.form.getFormId()).toBe(formId);
  });

  it('should receive the formCustom in the props', () => {
    const FormContainer = _createForm(formId, "formCustom")(Form);
    const tree = TestUtils.renderIntoDocument(
      <FormContainer
        dispatch={dispatch}
        formId={formId}
        init={jest.fn()}
      />
    );
    const form = TestUtils.findRenderedComponentWithType(tree, Form);
    expect(form.props.formCustom.getFormId()).toBe(formId);
  });
});

describe('form', () => {
  let fields, fieldNames, form;

  beforeEach(() => {
    fields = {};
    fieldNames = [];
    form = __createForm(fields, fieldNames)(formId, dispatch);
  });

  it('should dispatch an action to lock submission', () => {
    expect(form.lockSubmission()).toMatchSnapshot();
  });

  it('should dispatch an action to unlock submission', () => {
    expect(form.unlockSubmission()).toMatchSnapshot();
  });

  it('should dispatch an action to disable form', () => {
    expect(form.disable()).toMatchSnapshot();
  });

  it('should dispatch an action to enable form', () => {
    expect(form.enable()).toMatchSnapshot();
  });

  it('should dispatch an action to disable field', () => {
    expect(form.disableField(['foo'])).toMatchSnapshot();
    expect(form.disableField(['foo', 'bar'])).toMatchSnapshot();
  });

  it('should dispatch an action to enable field', () => {
    expect(form.enableField(['foo'])).toMatchSnapshot();
    expect(form.enableField(['foo', 'bar'])).toMatchSnapshot();
  });

  it('should dispatch an action to disable fields', () => {
    expect(form.disableFields([['foo'], ['bar']])).toMatchSnapshot();
    expect(form.disableFields([['foo', 'bar'], ['oof', 'rab']])).toMatchSnapshot();
  });

  it('should dispatch an action to enable fields', () => {
    expect(form.enableFields([['foo'], ['bar']])).toMatchSnapshot();
    expect(form.enableFields([['foo', 'bar'], ['oof', 'rab']])).toMatchSnapshot();
  });

  it('should dispatch an action to set value', () => {
    expect(form.setValue(['foo'], 'Hello')).toMatchSnapshot();
    expect(form.setValue(['foo', 'bar'], 'Hello World')).toMatchSnapshot();
  });

  it('should dispatch an action to set values', () => {
    expect(form.setValues([{
      fieldName: ['foo'],
      value: 'Hello',
    }, {
      fieldName: ['foo', 'bar'],
      value: 'Hello World',
    }])).toMatchSnapshot();
  });

  it('should dispatch an action to set default value', () => {
    expect(form.setDefaultValue(['foo'], 'Hello')).toMatchSnapshot();
    expect(form.setDefaultValue(['foo', 'bar'], 'Hello World')).toMatchSnapshot();
  });

  it('should dispatch an action to set default values', () => {
    expect(form.setDefaultValues([{
      fieldName: ['foo'],
      value: 'Hello',
    }, {
      fieldName: ['foo', 'bar'],
      value: 'Hello World',
    }])).toMatchSnapshot();
  });

  it('should dispatch an action to set error messages', () => {
    expect(form.setErrors(['foo'], ['Error 1', 'Error 2'])).toMatchSnapshot();
    expect(form.setErrors(['foo', 'bar'], ['Error 1', 'Error 2'])).toMatchSnapshot();
  });

  it('should dispatch an action to add error messages', () => {
    expect(form.addErrors(['foo'], ['Error 1', 'Error 2'])).toMatchSnapshot();
    expect(form.addErrors(['foo', 'bar'], ['Error 1', 'Error 2'])).toMatchSnapshot();
  });

  it('should dispatch an action to clear form', () => {
    expect(form.clearForm()).toMatchSnapshot();
  });

  it('should dispatch an action to clear error messages', () => {
    expect(form.clearErrors(['foo'])).toMatchSnapshot();
    expect(form.clearErrors(['foo', 'bar'])).toMatchSnapshot();
  });

  it('should register a field', () => {
    form.regField(['foo'], jest.fn());
    expect({
      fields,
      fieldNames,
    }).toMatchSnapshot();
  });

  it('should unregister a field', () => {
    form.regField(['foo'], jest.fn());
    form.regField(['bar'], jest.fn());
    form.regField(['foo', 'bar'], jest.fn());
    form.regField(['foo', 'bar', 'bar', 0], jest.fn());
    form.regField(['foo', 'bar', 0, 'bar'], jest.fn());

    const _previous = {
      fields: { ...fields },
      fieldNames: [...fieldNames],
    };
    form.unregField(['foo']);
    form.unregField(['foo']);
    expect({ _previous, current: { fields, fieldNames }}).toMatchSnapshot();

    form.unregField(['foo', 'bar']);
    expect({ _previous, current: { fields, fieldNames }}).toMatchSnapshot();

    form.unregField(['foo', 'bar', 0, 'bar']);
    expect({ _previous, current: { fields, fieldNames }}).toMatchSnapshot();
  });

  it('should validate all registered fields', async () => {
    form.regField(['foo', 'bar'], () => Promise.resolve({
      isValid: true,
      value: 'Value 1',
      name: ['foo', 'bar'],
    }));
    form.regField(['foo', 'oof', 0], () => Promise.resolve({
      isValid: true,
      value: 'Value 2',
      name: ['foo', 'oof', 0],
    }));
    expect(await form.validateAll()).toMatchSnapshot();
  });

  it('should validate all registered fields', async () => {
    form.regField([0, 'foo'], () => Promise.resolve({
      isValid: true,
      value: 'Value 1',
      name: [0, 'foo'],
    }));
    expect(await form.validateAll()).toMatchSnapshot();
  });

  it('should throw on validations error', async () => {
    form.regField(['foo', 'bar'], () => Promise.reject(new Error('Field is required')));
    try {
      await form.validateAll();
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });

  it('should submit the form with submission callback', async () => {
    form.regField(['foo', 'bar'], () => Promise.resolve({
      isValid: false,
      value: 'Value 1',
      name: ['foo', 'bar'],
    }));
    form.regField(['foo', 'oof', 0], () => Promise.resolve({
      isValid: true,
      value: 'Value 2',
      name: ['foo', 'oof', 0],
    }));
    await form.submit(result => {
      expect(result).toMatchSnapshot();
    });
  });

  it('should submit the form with submission and validation callback', async () => {
    form.regField(['foo', 'bar'], () => Promise.resolve({
      isValid: true,
      value: 'Value 1',
      name: ['foo', 'bar'],
    }));
    form.regField(['foo', 'oof', 0], () => Promise.resolve({
      isValid: true,
      value: 'Value 2',
      name: ['foo', 'oof', 0],
    }));
    const validationFunc1 = jest.fn();
    await form.submit(
      result => {
        expect(validationFunc1.mock.calls.length).toBe(1);
        expect(result).toMatchSnapshot();
      },
      validationFunc1
    );
    const validationFunc2 = jest.fn(() => Promise.resolve({
      isValid: false,
      errorMessages: ['Field must be a number'],
    }));
    await form.submit(
      result => {
        expect(validationFunc2.mock.calls.length).toBe(1);
        expect(result).toMatchSnapshot();
      },
      validationFunc2
    );
  });

  it('should throw on form submission error', async () => {
    form.regField(['foo', 'bar'], () => Promise.resolve({
      isValid: true,
      value: 'Hello World',
      name: ['foo', 'bar'],
    }));
    const submitFunc = jest.fn();
    try {
      await form.submit(
        submitFunc,
        () => Promise.reject(new Error('Form validation error'))
      );
    } catch (error) {
      expect(submitFunc.mock.calls.length).toBe(1);
      expect(error).toMatchSnapshot();
    }
  });

  it('should throw on form submission error', async () => {
    form.regField(['foo', 'bar'], () => Promise.reject(new Error('Form submission error')));
    try {
      await form.submit();
    } catch (error) {
      expect(error).toMatchSnapshot();
    }
  });
});
