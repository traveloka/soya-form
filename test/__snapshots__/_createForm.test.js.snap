// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form should dispatch an action to add error messages 1`] = `
Object {
  "errorMessages": Array [
    "foo",
  ],
  "formId": "simpleForm",
  "type": "form.addErrors",
}
`;

exports[`form should dispatch an action to add error messages 2`] = `
Object {
  "errorMessages": Array [
    "foo",
    "bar",
  ],
  "formId": "simpleForm",
  "type": "form.addErrors",
}
`;

exports[`form should dispatch an action to clear error messages 1`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": "foo",
    },
  ],
  "formId": "simpleForm",
  "type": "form.clearErrors",
}
`;

exports[`form should dispatch an action to clear error messages 2`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": "foo",
    },
    Object {
      "fieldName": "bar",
    },
  ],
  "formId": "simpleForm",
  "type": "form.clearErrors",
}
`;

exports[`form should dispatch an action to clear form 1`] = `
Object {
  "formId": "simpleForm",
  "type": "form.clearForm",
}
`;

exports[`form should dispatch an action to disable field 1`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": Array [
        "foo",
      ],
      "object": Object {
        "isEnabled": false,
      },
    },
  ],
  "formId": "simpleForm",
  "type": "form.mergeFields",
}
`;

exports[`form should dispatch an action to disable field 2`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": Array [
        "foo",
        "bar",
      ],
      "object": Object {
        "isEnabled": false,
      },
    },
  ],
  "formId": "simpleForm",
  "type": "form.mergeFields",
}
`;

exports[`form should dispatch an action to disable fields 1`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": Array [
        "foo",
      ],
      "object": Object {
        "isEnabled": false,
      },
    },
    Object {
      "fieldName": Array [
        "bar",
      ],
      "object": Object {
        "isEnabled": false,
      },
    },
  ],
  "formId": "simpleForm",
  "type": "form.mergeFields",
}
`;

exports[`form should dispatch an action to disable fields 2`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": Array [
        "foo",
        "bar",
      ],
      "object": Object {
        "isEnabled": false,
      },
    },
    Object {
      "fieldName": Array [
        "oof",
        "rab",
      ],
      "object": Object {
        "isEnabled": false,
      },
    },
  ],
  "formId": "simpleForm",
  "type": "form.mergeFields",
}
`;

exports[`form should dispatch an action to disable form 1`] = `
Object {
  "formId": "simpleForm",
  "isEnabled": false,
  "type": "form.setEnabledState",
}
`;

exports[`form should dispatch an action to enable field 1`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": Array [
        "foo",
      ],
      "object": Object {
        "isEnabled": true,
      },
    },
  ],
  "formId": "simpleForm",
  "type": "form.mergeFields",
}
`;

exports[`form should dispatch an action to enable field 2`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": Array [
        "foo",
        "bar",
      ],
      "object": Object {
        "isEnabled": true,
      },
    },
  ],
  "formId": "simpleForm",
  "type": "form.mergeFields",
}
`;

exports[`form should dispatch an action to enable fields 1`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": Array [
        "foo",
      ],
      "object": Object {
        "isEnabled": true,
      },
    },
    Object {
      "fieldName": Array [
        "bar",
      ],
      "object": Object {
        "isEnabled": true,
      },
    },
  ],
  "formId": "simpleForm",
  "type": "form.mergeFields",
}
`;

exports[`form should dispatch an action to enable fields 2`] = `
Object {
  "fields": Array [
    Object {
      "fieldName": Array [
        "foo",
        "bar",
      ],
      "object": Object {
        "isEnabled": true,
      },
    },
    Object {
      "fieldName": Array [
        "oof",
        "rab",
      ],
      "object": Object {
        "isEnabled": true,
      },
    },
  ],
  "formId": "simpleForm",
  "type": "form.mergeFields",
}
`;

exports[`form should dispatch an action to enable form 1`] = `
Object {
  "formId": "simpleForm",
  "isEnabled": true,
  "type": "form.setEnabledState",
}
`;

exports[`form should dispatch an action to lock submission 1`] = `
Object {
  "formId": "simpleForm",
  "isSubmitting": true,
  "type": "form.setIsSubmitting",
}
`;

exports[`form should dispatch an action to set default value 1`] = `
Object {
  "fieldName": Array [
    "foo",
  ],
  "formId": "simpleForm",
  "type": "form.setDefaultValue",
  "value": "Hello",
}
`;

exports[`form should dispatch an action to set default value 2`] = `
Object {
  "fieldName": Array [
    "foo",
    "bar",
  ],
  "formId": "simpleForm",
  "type": "form.setDefaultValue",
  "value": "Hello World",
}
`;

exports[`form should dispatch an action to set default values 1`] = `
Object {
  "formId": "simpleForm",
  "type": "form.setDefaultValues",
  "values": Array [
    Object {
      "fieldName": Array [
        "foo",
      ],
      "value": "Hello",
    },
    Object {
      "fieldName": Array [
        "foo",
        "bar",
      ],
      "value": "Hello World",
    },
  ],
}
`;

exports[`form should dispatch an action to set error messages 1`] = `
Object {
  "errorMessages": Array [
    "Error 1",
    "Error 2",
  ],
  "fieldName": Array [
    "foo",
  ],
  "formId": "simpleForm",
  "type": "form.setErrors",
}
`;

exports[`form should dispatch an action to set error messages 2`] = `
Object {
  "errorMessages": Array [
    "Error 1",
    "Error 2",
  ],
  "fieldName": Array [
    "foo",
    "bar",
  ],
  "formId": "simpleForm",
  "type": "form.setErrors",
}
`;

exports[`form should dispatch an action to set value 1`] = `
Object {
  "fieldName": Array [
    "foo",
  ],
  "formId": "simpleForm",
  "type": "form.setValue",
  "value": "Hello",
}
`;

exports[`form should dispatch an action to set value 2`] = `
Object {
  "fieldName": Array [
    "foo",
    "bar",
  ],
  "formId": "simpleForm",
  "type": "form.setValue",
  "value": "Hello World",
}
`;

exports[`form should dispatch an action to set values 1`] = `
Object {
  "formId": "simpleForm",
  "type": "form.setValues",
  "values": Array [
    Object {
      "fieldName": Array [
        "foo",
      ],
      "value": "Hello",
    },
    Object {
      "fieldName": Array [
        "foo",
        "bar",
      ],
      "value": "Hello World",
    },
  ],
}
`;

exports[`form should dispatch an action to unlock submission 1`] = `
Object {
  "formId": "simpleForm",
  "isSubmitting": false,
  "type": "form.setIsSubmitting",
}
`;

exports[`form should register a field 1`] = `
Object {
  "fieldNames": Array [
    Array [
      "foo",
    ],
  ],
  "fields": Object {
    "foo": Object {
      "validateAll": [MockFunction],
    },
  },
}
`;

exports[`form should submit the form with submission and validation callback 1`] = `
Object {
  "isValid": true,
  "values": Object {
    "foo": Object {
      "bar": "Value 1",
      "oof": Array [
        "Value 2",
      ],
    },
  },
}
`;

exports[`form should submit the form with submission and validation callback 2`] = `
Object {
  "isValid": false,
  "values": Object {
    "foo": Object {
      "bar": "Value 1",
      "oof": Array [
        "Value 2",
      ],
    },
  },
}
`;

exports[`form should submit the form with submission callback 1`] = `
Object {
  "isValid": false,
  "values": Object {
    "foo": Object {
      "bar": "Value 1",
      "oof": Array [
        "Value 2",
      ],
    },
  },
}
`;

exports[`form should throw on form submission error 1`] = `[Error: Form validation error]`;

exports[`form should throw on form submission error 2`] = `[Error: Form submission error]`;

exports[`form should throw on validations error 1`] = `[Error: Field is required]`;

exports[`form should unregister a field 1`] = `
Object {
  "_previous": Object {
    "fieldNames": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "foo",
        "bar",
      ],
      Array [
        "foo",
        "bar",
        "bar",
        0,
      ],
      Array [
        "foo",
        "bar",
        0,
        "bar",
      ],
    ],
    "fields": Object {
      "bar": Object {
        "validateAll": [MockFunction],
      },
      "foo": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,0,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,bar,0": Object {
        "validateAll": [MockFunction],
      },
    },
  },
  "current": Object {
    "fieldNames": Array [
      Array [
        "bar",
      ],
      Array [
        "foo",
        "bar",
      ],
      Array [
        "foo",
        "bar",
        "bar",
        0,
      ],
      Array [
        "foo",
        "bar",
        0,
        "bar",
      ],
    ],
    "fields": Object {
      "bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,0,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,bar,0": Object {
        "validateAll": [MockFunction],
      },
    },
  },
}
`;

exports[`form should unregister a field 2`] = `
Object {
  "_previous": Object {
    "fieldNames": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "foo",
        "bar",
      ],
      Array [
        "foo",
        "bar",
        "bar",
        0,
      ],
      Array [
        "foo",
        "bar",
        0,
        "bar",
      ],
    ],
    "fields": Object {
      "bar": Object {
        "validateAll": [MockFunction],
      },
      "foo": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,0,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,bar,0": Object {
        "validateAll": [MockFunction],
      },
    },
  },
  "current": Object {
    "fieldNames": Array [
      Array [
        "bar",
      ],
      Array [
        "foo",
        "bar",
        "bar",
        0,
      ],
      Array [
        "foo",
        "bar",
        0,
        "bar",
      ],
    ],
    "fields": Object {
      "bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,0,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,bar,0": Object {
        "validateAll": [MockFunction],
      },
    },
  },
}
`;

exports[`form should unregister a field 3`] = `
Object {
  "_previous": Object {
    "fieldNames": Array [
      Array [
        "foo",
      ],
      Array [
        "bar",
      ],
      Array [
        "foo",
        "bar",
      ],
      Array [
        "foo",
        "bar",
        "bar",
        0,
      ],
      Array [
        "foo",
        "bar",
        0,
        "bar",
      ],
    ],
    "fields": Object {
      "bar": Object {
        "validateAll": [MockFunction],
      },
      "foo": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,0,bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,bar,0": Object {
        "validateAll": [MockFunction],
      },
    },
  },
  "current": Object {
    "fieldNames": Array [
      Array [
        "bar",
      ],
      Array [
        "foo",
        "bar",
        "bar",
        0,
      ],
    ],
    "fields": Object {
      "bar": Object {
        "validateAll": [MockFunction],
      },
      "foo,bar,bar,0": Object {
        "validateAll": [MockFunction],
      },
    },
  },
}
`;

exports[`form should validate all registered fields 1`] = `
Object {
  "isValid": true,
  "values": Object {
    "foo": Object {
      "bar": "Value 1",
      "oof": Array [
        "Value 2",
      ],
    },
  },
}
`;

exports[`form should validate all registered fields 2`] = `
Object {
  "isValid": true,
  "values": Array [
    Object {
      "foo": "Value 1",
    },
  ],
}
`;
